### PR TITLE
feat(actions)!: add full sync test

### PR DESCRIPTION
**Pull request by gustavovalverde**
## Motivation
Most of the motivation for this PR is available at #1592 

### Specifications
- We need to run a full sync test on every PR before merging to main
- Preferably this test should run at least once before starting the merge process
- Multiple tests must be done to accomplish the best balance between performance and cost in GCP

### Designs
- Use a custom `entrypoint.sh` to overcome some limitations from GCP docker parser

## Solution
- Execute the full sync test using an environment variable
- WIP

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

**Comment by teor2345** [comment]
We got to 87%, then the job timed out at 6 hours:

> Feb 22 02:03:48.658  INFO ***net="Main"***: zebrad::commands::start: estimated progress to chain tip sync_percent=87.748 % current_height=Height(1382111) remaining_sync_blocks=192980 time_since_last_state_block=PT0S

https://github.com/ZcashFoundation/zebra/runs/5279457879?check_suite_focus=true#step:7:8774

**Comment by gustavovalverde** [comment]
Yeah, after 50% it started to get slow. I'll try a bigger machine to see if there's an impact there. Any other recommendations (including changes in the configuration file) are welcome @teor2345

**Comment by gustavovalverde** [comment]
Still running here: https://cloudlogging.app.goo.gl/RxUMH5twUULrijba9

**Comment by teor2345** [comment]
> Yeah, after 50% it started to get slow. I'll try a bigger machine to see if there's an impact there. Any other recommendations (including changes in the configuration file) are welcome @teor2345

I'd suggest:
- try the biggest machine possible, just for now
- setting `checkpoint_sync = true`
- update the mainnet and pre-NU5 testnet checkpoint lists (part of #2368)
- #3603, if we're seeing lots of those logs
- work out if the test is slowed down by network, CPU, RAM, or disk limits

**Comment by teor2345** [comment]
I can do the checkpoint lists now, they're the quick part of that ticket.

**Comment by gustavovalverde** [comment]
Could it be something related to the amount of data (blocks) we're getting to be processed? At the beginning the network and disks had a big spike, and ended a bit after the sync call (just were the red rectangle ends), which makes me thing the disk shouldn't be the ones halting the process, as they can handled the spikes nicely.

image

I'll increase the machine type anyways.

**Comment by teor2345** [comment]
I updated the checkpoint lists in:
- #3606

If we use that PR and `checkpoint_sync = true`, the sync should be a lot faster.

**Comment by teor2345** [comment]
> Could it be something related to the amount of data (blocks) we're getting to be processed?

It could be a concurrency issue. We can limit the number of blocks in the queue by fixing:
- #3603

But let's try the checkpoints first, they should speed things up a lot.

**Comment by teor2345** [comment]
The full sync test failed due to a request error, so I tried to re-run it. But it seems like build images is also failing with the same error.

**Comment by teor2345** [comment]
It seems like we're not failing on some full sync failures.

This job passed, despite this error:
```
+ exec cargo test --locked --release --features enable-sentry --test acceptance -- --nocapture --ignored full_sync_mainnet
   Compiling zebrad v1.0.0-beta.5 (/app/zebrad)
error: failed to run custom build command for `zebrad v1.0.0-beta.5 (/app/zebrad)`

Caused by:
  process didn't exit successfully: `/app/target/release/build/zebrad-d50ef2b2564080da/build-script-build` (signal: 4, SIGILL: illegal instruction)
```
https://github.com/ZcashFoundation/zebra/runs/5325691695?check_suite_focus=true#step:7:117

**Comment by teor2345** [comment]
I'm currently running `c2d-standard-16` with:
- `-C opt-level=2` - 100% at 5:59 - #3582 (finished, but failed due to a test harness bug)
- `-C opt-level=3` - 100% at 5:49 - #3635 
- `-C opt-level=z` - ~88% at 6:00 - #3636
- `-C opt-level=s` - 100% at 6:00 - #3639 (almost finished, 2 blocks to go)

**Comment by teor2345** [comment]
Zebra succeeded in 5:59, but the test failed due to a test harness bug.
I pushed a fix in:
- #3643

I'll cherry-pick the fix and the best optimization level to this PR.

**Comment by teor2345** [comment]
The fastest sync was `-C opt-level=3` - 100% at 5:49 - https://github.com/ZcashFoundation/zebra/pull/3635
The build was also about 15 minutes faster.

I'm currently running `-C opt-level=3` and the test harness fix with:
- gossip limit ` 20` - 100% at 5:50 - #3582 - no change to the limit
- gossip limit `  5` - 99.992% at 6:00 - #3644
- gossip limit `100` - 99.994% at 6:00 - #3645

Other quick fixes we could try are the concurrent block download limit, and the sync lookahead limit.

**Comment by teor2345** [comment]
The fastest sync was gossip limit `20` - 100% at 5:50 - #3582 (no change to the limit).

I'm currently running that config with:
- concurrent sync blocks `30` - 100% at 5:54 - #3651
- concurrent sync blocks `40` - 100% at 5:49 - #3652
- concurrent sync blocks `60` - 100% at 5:48 - #3653

The default is `50` concurrent sync blocks, which takes about 5:52.

The other quick fix we could try is the sync lookahead limit.

If that doesn't work, we'll need to do some detailed analysis to:
- #3375

**Comment by teor2345** [comment]
There was no significant difference between 40, 50, and 60 concurrent blocks, so I'll just leave it at 50 for now.

**Comment by teor2345** [comment]
I'm currently running this branch with:
- lookahead limit `1200` - 100% at 5:58 - #3661
- lookahead limit `2000` - 100% at 5:30 - #3662 - the default
- lookahead limit `2800` - 100% at 5:47 - #3663

PR #3662 didn't change any code, but it was half an hour faster than the original. This could have happened due to the switch from Rust 1.58 to 1.59.

If this performance is consistent, we might not need to make any more performance improvements right now.

File: docker/entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+checkpoint_sync = ${CHECKPOINT_SYNC}
+[metrics]
+endpoint_addr = 0.0.0.0:9999
+[network]
+network = ${NETWORK}
+[state]
+cache_dir = /zebrad-cache
+[tracing]
+force_use_color = true
+endpoint_addr = 0.0.0.0:3000" > /app/zebrad.toml
```
Comment (teor2345) [inline docker/entrypoint.sh:17]: Security Question: Does binding to `0.0.0.0` inside a docker container expose the port to the whole internet?

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -90,12 +90,307 @@ jobs:
           cache-from: type=gha
           cache-to: type=gha,mode=max
 
-  # Run all the zebra tests, including tests that are ignored by default
-  test-all:
-    name: Test all
+  # # Run all the zebra tests, including tests that are ignored by default
```
Comment (teor2345) [inline .github/workflows/test.yml]: Blocker: Most of our tests will be disabled if we merge this PR.

This looks like a left over development hack?

**Review by teor2345 — DISMISSED** [review]
I don't think this PR is ready to merge yet, because it disables a lot of existing tests.

File: docker/entrypoint.sh (hunk 1 above)
Comment (gustavovalverde) [reply docker/entrypoint.sh:17]: It expose it to the host. If you configure the host with forwarding, then it also forwards to the internet, but that's not the default behavior.

File: .github/workflows/test.yml (hunk 2 above)
Comment (gustavovalverde) [reply .github/workflows/test.yml]: Yes, this was a development hack so we didn't had to wait for the whole process. If we're done with testing the timing, then I'll finalize this PR for merging.

Comment (teor2345) [reply .github/workflows/test.yml]: I've tried all the quick tweaks I could think of.

If we want more speedups, we'll need to do analysis like:
- #3375

Comment (gustavovalverde) [reply .github/workflows/test.yml]: I think we can do that afterwards. From my PoV there's more benefit on merging this PR ASAP and solving issues in the road than actually waiting for 30 minutes or 1h less in the test.

Comment (teor2345) [reply .github/workflows/test.yml]: Yes, I agree.

**Review by teor2345 — APPROVED** [review]
I think we're all done here, thanks for your patience working through all the sync issues.

**Pull request merged**